src/pyhf/events.py

```diff
@@ -24,19 +23,66 @@ def noop(*args, **kwargs):
     pass
 
 
-class WeakList(list):
-    def append(self, item):
-        list.append(self, weakref.WeakMethod(item, self.remove))
+class Callables:
+    def __init__(self):
+        self._callbacks = []
+
+    @property
+    def callbacks(self):
+        """
+        Get the current list of living callbacks.
+        """
+        self._flush()
+        return self._callbacks
+
+    def append(self, callback):
+        """
+        Append a new bound method as a callback to the list of callables.
+        """
+        try:
+            # methods
+            callback_ref = weakref.ref(callback.__func__), weakref.ref(
+                callback.__self__
+            )
+        except AttributeError:
+            callback_ref = weakref.ref(callback), None
+        self._callbacks.append(callback_ref)
```

Comment:
```suggestion
        self._callbacks.append(
            (weakref.ref(callback.__func__), weakref.ref(callback.__self__))
        )
```

Reply:
Ah a very easy counter example that I missed is the following commented out example in the file:

```python
>>> import pyhf
>>> @pyhf.events.subscribe('myevent')
... def test(a,b):
...     print(a+b)
... 
>>> pyhf.events.trigger("myevent")(1,2)
3
```

If you tried it with things removed like I suggested above you'd get the following error

```python
>>> import pyhf
>>> @pyhf.events.subscribe('myevent')
... def test(a,b):
...     print(a+b)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 91, in __decorator
    __events.setdefault(event, Callables()).append(func)
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/events.py", line 43, in append
    (weakref.ref(callback.__func__), weakref.ref(callback.__self__))
AttributeError: 'function' object has no attribute '__func__'
```

My bad.